plugins/database/postgresql/postgresql.go

```diff
@@ -68,19 +72,46 @@ func new() *PostgreSQL {
 
 type PostgreSQL struct {
 	*connutil.SQLConnectionProducer
+
+	usernameProducer template.StringTemplate
 }
 
 func (p *PostgreSQL) Initialize(ctx context.Context, req dbplugin.InitializeRequest) (dbplugin.InitializeResponse, error) {
 	newConf, err := p.SQLConnectionProducer.Init(ctx, req.Config, req.VerifyConnection)
 	if err != nil {
 		return dbplugin.InitializeResponse{}, err
 	}
+
+	usernameTemplate := getString(req.Config, "username_template")
+	if usernameTemplate == "" {
+		usernameTemplate = defaultUserNameTemplate
+	}
+
+	up, err := template.NewTemplate(template.Template(usernameTemplate))
+	if err != nil {
+		return dbplugin.InitializeResponse{}, fmt.Errorf("unable to initialize username template: %w", err)
+	}
+	p.usernameProducer = up
+
+	_, err = p.usernameProducer.Generate(dbplugin.UsernameMetadata{})
+	if err != nil {
+		return dbplugin.InitializeResponse{}, fmt.Errorf("invalid username template - did you reference a field that isn't available? : %w", err)
+	}
+
 	resp := dbplugin.InitializeResponse{
 		Config: newConf,
 	}
 	return resp, nil
 }
 
+func getString(m map[string]interface{}, key string) string {
+	var result string
+	if err := mapstructure.Decode(m[key], &result); err != nil {
```

Comment:
Can we access the map and do an ok check + type assertion here? `mapstructure.Decode` ends up calling `reflect` funcs which might be a bit heavy-handed for this.

Reply:
Updated the `GetString` function to not use mapstructure. I also moved it into the `strutil` package as I find I have needed to use it in a bunch of places.